x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/common/validation/SourceDestValidator.java

```diff
@@ -0,0 +1,513 @@
+     */
+
+    public static ValidationException validate(
+        ClusterState clusterState,
+        IndexNameExpressionResolver indexNameExpressionResolver,
+        RemoteClusterService remoteClusterService,
+        RemoteClusterLicenseChecker remoteClusterLicenseChecker,
+        String[] source,
+        String dest,
+        String nodeName,
+        String license,
+        boolean shouldDefer
+    ) {
+        Context context = new Context(
+            clusterState,
+            indexNameExpressionResolver,
+            remoteClusterService,
+            remoteClusterLicenseChecker,
+            source,
+            dest,
+            nodeName,
+            license
+        );
+
+        // validation exception might gets thrown if validation stops early
+        try {
+            for (SourceDestValidation validation : START_AND_PUT_VALIDATIONS) {
+                if (shouldDefer && validation.isDeferrable()) {
+                    continue;
+                }
+                validation.validate(context);
+            }
+
+        } catch (ValidationException e) {}
+
+        return context.getValidationException();
+    }
+
+    static class SourceMissingValidation implements SourceDestValidation {
+
+        @Override
+        public boolean isDeferrable() {
+            return true;
+        }
+
+        @Override
+        public void validate(Context context) {
+            // boolean remoteIndexesEmpty = context.isRemoteSearchEnabled() ? context.resolveRemoteSource().isEmpty() : true;
+
+            try {
+                // non-trivia: if source contains a wildcard index, which does not resolve to a concrete index
+                // the resolved indices might be empty, but we can check if source contained something, this works because
+                // of no wildcard index is involved the resolve would have thrown an exception
+                if (context.resolveSource().isEmpty() && context.resolveRemoteSource().isEmpty() && context.getSource().length == 0) {
+                    context.addValidationError(SOURCE_INDEX_MISSING, true, Strings.arrayToCommaDelimitedString(context.getSource()));
+                }
+            } catch (IndexNotFoundException e) {
+                context.addValidationError(e.getMessage(), true);
+            }
+        }
+    }
+
+    static class RemoteSourceEnabledAndRemoteLicenseValidation implements SourceDestValidation {
+        @Override
+        public boolean isDeferrable() {
+            return true;
+        }
+
+        @Override
+        public void validate(Context context) {
+            if (context.resolveRemoteSource().isEmpty()) {
+                return;
+            }
+
+            List<String> remoteIndices = new ArrayList<>(context.resolveRemoteSource());
+            // we can only check this node at the moment, clusters with mixed CCS enabled/disabled nodes are not supported,
+            // see gh#50033
```

Comment:
NOTE (not for this PR): I think we should still do this validation in the executor (remote search + license check). Should be possible

Reply:
#50033 has some analysis of this problem, it requires some work to get this working.

Note that we do validation at the start of the task, however the problem is `cluster.remote` being a node setting which you can not get via cluster state. #50033 has the details, if you have ideas/comments please put it there.